### PR TITLE
feat: impl dal

### PR DESCRIPTION
**Pull request by killagu**
##### Checklist

- [ ] `npm test` passes
- [ ] tests and/or benchmarks are included
- [ ] documentation is changed or added
- [ ] commit message follows commit guidelines

## Summary by CodeRabbit

- **New Features**
	- Introduced database abstraction layer (DAL) functionality, including decorators for columns, tables, indexes, and data source qualification.
	- Added extensive enumeration support for database-related properties such as column formats, types, compression types, and more.
	- Launched new classes for modeling database columns, indexes, and tables.
	- Provided utilities for managing column, index, and table information in DAL entities.
	- Enhanced DAL documentation with usage guidance and reference to additional resources.
- **Documentation**
	- Updated documentation to include guidance on using new DAL decorators and runtime features.
- **Tests**
	- Expanded test coverage to include new DAL functionalities and decorators.
- **Chores**
	- Updated `.gitignore` to exclude `.node` files.

### CODE REVIEW

File: core/dal-decorator/src/decorator/Column.ts [hunk 1]
```diff
@@ -0,0 +1,273 @@
...
+  canNull?: boolean;
+  comment?: string;
+  visible?: boolean;
+  autoIncrement?: boolean;
+  uniqueKey?: boolean;
+  primaryKey?: boolean;
+  collate?: string;
+  columnFormat?: ColumnFormat;
+  engineAttribute?: string;
+  secondaryEngineAttribute?: string;
```
Comment (fengmk2) [inline core/dal-decorator/src/decorator/Column.ts:19]: characterSet 是否也需要？

Comment (fengmk2) [reply core/dal-decorator/src/decorator/Column.ts:19]: 哦，看到是放到具体的 VarCharParams 里面了。

File: core/dal-decorator/src/util/TableInfoUtil.ts [hunk 2]
```diff
@@ -0,0 +1,23 @@
+import { EggProtoImplClass, MetadataUtil } from '@eggjs/core-decorator';
+import { TableParams } from '../decorator/Table';
+
+export const DAL_IS_TABLE = Symbol.for('EggPrototype#dalIsTable');
+export const DAL_TABLE_PARAMS = Symbol.for('EggPrototype#dalTableParams');
```
Comment (fengmk2) [inline core/dal-decorator/src/util/TableInfoUtil.ts:5]: 是不是可以直接使用 Symbol 而不是 Symbol for？

File: plugin/dal/README.md [hunk 3]
```diff
@@ -0,0 +1,216 @@
...
+export class FooController {
+  @Middleware(methodCount)
+  async hello(@Context() ctx: EggContext) {
+  }
+}
+```
+
+### HTTP 注解
+
+#### HTTPController/HTTPMethod
```
Comment (fengmk2) [inline plugin/dal/README.md:90]: readme 内容改改

File: plugin/dal/test/fixtures/apps/dal-app/modules/dal/Foo.ts [hunk 4]
```diff
@@ -0,0 +1,292 @@
...
+  Line, MultiLine, MultiPoint, MultiPolygon, Point, Polygon,
+  Table,
+} from '@eggjs/dal-decorator';
+import { IndexStoreType } from '@eggjs/dal-decorator/src/enum/IndexStoreType';
+
+@Table({
+  name: 'egg_foo',
+  comment: 'foo table',
+  characterSet: 'utf8mb4',
+  collate: 'utf8mb4_unicode_ci',
```
Comment (fengmk2) [inline plugin/dal/test/fixtures/apps/dal-app/modules/dal/Foo.ts:17]: characterSet 和 collate 这些默认值是否有一个地方可以统一配置？要不然每个 class 都得写。

File: plugin/dal/lib/MysqlDataSourceManager.ts [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+  static instance = new MysqlDataSourceManager();
+
+  readonly dataSources: Map<string, Map<string, MysqlDataSource>>;
+
+  constructor() {
+    this.dataSources = new Map();
+  }
+
+  get(moduleName: string, dataSourceName: string): MysqlDataSource | undefined {
+    return this.dataSources.get(moduleName)?.get(dataSourceName);
```
Comment (fengmk2) [inline plugin/dal/lib/MysqlDataSourceManager.ts:13]: 所以每个 module 之间的 datasource 是完全隔离不会有连接复用之类的对么？

File: plugin/dal/test/fixtures/apps/dal-app/modules/dal/Foo.ts (hunk 4 above)
Comment (killagu) [reply plugin/dal/test/fixtures/apps/dal-app/modules/dal/Foo.ts:17]: 这会默认使用数据库的设置，不需要每个 class 配置的。是可选字段。

File: plugin/dal/lib/MysqlDataSourceManager.ts (hunk 5 above)
Comment (killagu) [reply plugin/dal/lib/MysqlDataSourceManager.ts:13]: dataSource 应该是独立的库了，所以 module 内的 dataSource 是需要独立的。

你是指一个 dataSource 在应用内的多个 module 间共享吗？的确有可能会导致连接数翻倍的情况，我来优化一下，如果 dataSource 的配置完全一样，复用同一个 dataSource。

Comment (fengmk2) [reply plugin/dal/lib/MysqlDataSourceManager.ts:13]: 对 +1

File: core/dal-decorator/src/model/TableModel.ts [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+    assert(TableInfoUtil.getIsTable(clazz as EggProtoImplClass), `${name} is not Table`);
+    assert(columnTypeMap, `${name} has no columns`);
+    const columns: Array<ColumnModel> = [];
+    const indices: Array<IndexModel> = [];
+    for (const [ property, columnType ] of columnTypeMap?.entries()) {
+      const columnParam = columnInfoMap?.get(property);
+      columns.push(ColumnModel.build(property, columnType, columnParam));
+    }
+
+    // TODO 补充 PRIMARY 主键
```
[the remark this answers is not in the document]
Comment (fengmk2) [reply core/dal-decorator/src/model/TableModel.ts:118]: @coderabbitai yes please help.

File: core/dal-runtime/src/BaseSqlMap.ts [hunk 7]
```diff
@@ -0,0 +1,324 @@
...
+        type: SqlType.SELECT,
+        sql,
+      });
+      sqlMaps.push({
+        name: `findOneBy${tempName}`,
+        type: SqlType.SELECT,
+        sql: `${sql} LIMIT 0, 1`,
+      });
+    }
+    return sqlMaps;
```
[the remark this answers is not in the document]
Comment (fengmk2) [reply core/dal-runtime/src/BaseSqlMap.ts:84]: @coderabbitai yes please

File: core/dal-runtime/src/SqlGenerator.ts [hunk 8]
```diff
@@ -0,0 +1,380 @@
+import {
+  BaseSpatialParams,
+  ColumnModel,
+  ColumnType,
+  ColumnTypeParams,
+  IndexModel, IndexType,
+  TableModel,
+} from '@eggjs/dal-decorator';
+
+// TODO diff 实现
```
[the remark this answers is not in the document]
Comment (killagu) [reply core/dal-runtime/src/SqlGenerator.ts:10]: Open a follow-up GitHub issue for this discussion.

**Comment by killagu** [comment]
@coderabbitai pause

**Comment by fengmk2** [comment]
+1

**Pull request merged**